### DISCUSSION TITLE
GraphQL API Deployment

### DISCUSSION THREAD

**Discussion by AmeyGohil** [sanity-io/sanity#3344] [Sanity Studio Support] [answered]
What's the comand for deploying graphql APIs to different workspaces in V3?
In V2, it is like `sanity graphql deploy --dataset staging` but this doesn't seem to work now.
And I couldn't find any relevant documentation for this in the [migration docs](https://beta.sanity.io/docs/platform/studio/v2-to-v3).

**Comment by p10e** [comment]
Hi @AmeyGohil! As you suggested, this seems to be missing from our documentation indeed - I will make a note for us to add it.

You should be able to deploy your GraphQL APIs by specifying the different workspaces in your `sanity.cli.ts` file. For example:

```
// sanity.cli.ts
import {createCliConfig} from 'sanity/cli'

export default createCliConfig({
  api: {
    projectId: <PROJECT_ID>,
    dataset: <DATASET_NAME>,
  },
  graphql: [
    {
      workspace: 'default',
      source: 'default', // Optional. Name of source containing the schema to deploy, within the configured workspace.
      tag: 'default', // Optional. API tag for this API - allows deploying multiple different APIs to a single dataset.
      playground: true, // Optional. Deploy a "GraphQL Playground" to the API url.
      generation: 'gen3', // Optional. Generation of API to auto-generate from schema. New APIs should use the latest (`gen3`).
      nonNullDocumentFields: false, // Optional. Define document interface fields (`_id`, `_type` etc) as non-nullable.
    },
    {
      workspace: 'staging',
    },
  ],
})
```

Please note that specifying multiple workspaces means GraphQL APIs will be deployed for each of them when you run the `sanity graphql deploy` command.

Hope this helps!

**Reply by AmeyGohil** [reply]
Thanks @p10e ! This works!

I have a follow-up question - What command should I run if I want to run gql deployment only for a particular workspace among those specified in `sanity.cli.ts`?
Of course, for example, I can change `sanity.cli.ts` to have only staging workspace and then run `sanity graphql deploy`, and repeat same for the production workspace but doing this can get quite tedious.

**Reply by p10e** [reply]
Great question, @AmeyGohil. I believe there is no such option available yet, but I can imagine us adding a flag like `--workspace <workspace>` the way we currently support this for a specific dataset.

For now, you may have to limit your entries to a single workspace in `sanity.cli.ts` if you want to expressly deploy a GraphQL API for just that workspace. I agree that this is not ideal and have forwarded the suggestion above to the team.

**Reply by snorrees** [reply]
Also remember that sanity.cli.js or ts is code! You could add your own filters to the workspace array before passing it to the create function.

**Reply by AmeyGohil** [reply]
Yes @snorrees, but I wanted to control the workspace via cli command, by some flag as @p10e mentioned earlier (like `--workspace <workspace>`). I am not sure how this can be via `sanity.ci.ts`.

**Comment by snorrees** [comment]
@AmeyGohil I realize I was a bit hasty in my answer, and didnt provide an example! Sorry about that. 
Here it is:

```ts
// sanity.cli.ts
import { createCliConfig } from "sanity/cli";

/**
 * Adds support for --workspace=<workspaceId>
 *
 * Can be provided multiple times to include more workspaces
 */
function getWorkspaceFilter(): string[] {
  // get at the npm arguments -> clean it up a bit along the way
  const args = (process.env.npm_lifecycle_script ?? "")
    .split(" ")
    .map((arg) => arg.replace(/['"]/g, ""));

  return args
    .filter((arg) => arg.startsWith("--workspace="))
    .map((arg) => arg.split("=")[1]);
}

const workspaceFilter = getWorkspaceFilter();

/* examples commands:

# Deploy all workspaces
npx sanity graphql deploy

# Deploy single workspace
npx sanity graphql deploy --workspace=staging

# Deploy multiple workspace
npx sanity graphql deploy --workspace=staging --workspace=preprod

*/
export default createCliConfig({
  api: {
    projectId: "<PROJECT_ID>",
    dataset: "<DATASET_NAME>",
  },
  graphql: [
    { workspace: "default" },
    { workspace: "staging" },
    { workspace: "preprod" },
  ].filter((gqlConfig) =>
    workspaceFilter.length
      ? workspaceFilter.includes(gqlConfig.workspace)
      : true
  ),
});
```

This adds a bit of code to get the original CLI command, parses it and extracts filters for workspace.
Essentially we are extending the cli command surface with what we need it to do 😃 Hopefully this is helpful! 

With this approach you can make the cli do anything with your config object really, based on any command line argument you desire ;)

**Reply by AmeyGohil** [reply]
Thanks a lot @snorrees! This really helps!

**Comment by rexxars** [comment]
Thanks for raising, @AmeyGohil - we are working on a native solution for this - adding a `--api` flag and allowing each API to have an `id` you can target. PR is here: https://github.com/sanity-io/sanity/pull/3366

Might not see a release before next week, however.

**Comment by rexxars** [comment, marked as answer]
v3.0.0-dev-preview.10 is out and includes the `--api` flag 👍 

You can now set an `id` on GraphQL APIs in your `sanity.cli.ts`:

```ts
// sanity.cli.ts
import {createCliConfig} from 'sanity/cli'

export default createCliConfig({
  api: {
    projectId: <PROJECT_ID>,
    dataset: <DATASET_NAME>,
  },
  graphql: [
    {
      id: 'production',
      workspace: 'default',
    },
    {
      id: 'staging',
      workspace: 'staging',
    },
  ],
})
```

The `id` is used in combination with the `--api` flag:

```
sanity graphql deploy --api staging
```

**Reply by AmeyGohil** [reply]
Thanks a lot! How do I update this in my setup? Would `sanity upgrade` do it or is there any other command for this?

**Reply by rexxars** [reply]
Right now you'll want to manually upgrade the version number for `sanity` in `package.json` and run `npm install`/`yarn`/whatever package manager you use.
